engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,671 @@
+                    resultModifiedColumnSet[ii] =
+                            result.newModifiedColumnSet(addNames);
+                    final ModifiedColumnSet.Transformer transformer =
+                            input.newModifiedColumnSetTransformer(result, addNames);
+
+                    final MultiJoinListenerRecorder listenerRecorder =
+                            new MultiJoinListenerRecorder("multiJoin(" + ii + ")", input, result, keySources,
+                                    sourceKeyModifiedColumnSet, transformer, ii);
+                    listenerRecorder.setMergedListener(mergedListener);
+                    input.addUpdateListener(listenerRecorder);
+                    listenerRecorders.add(listenerRecorder);
+                }
+            }
+
+            result.addParentReference(mergedListener);
+        }
+
+        return result;
+    }
+
+    @NotNull
+    private static MultiJoinInput maybePopulateColumnsToAdd(
+            @NotNull final Collection<String> keyColumnNames,
+            @NotNull final MultiJoinInput[] multiJoinInputs, final int tableNumber) {
+        MultiJoinInput joinInput = multiJoinInputs[tableNumber];
+        JoinAddition[] columnsToAdd = joinInput.columnsToAdd();
+        if (columnsToAdd.length == 0) {
+            // create them on the fly from the table
+            final JoinAddition[] newColumnsToAdd = joinInput.inputTable().getDefinition().getColumnNames().stream()
+                    .filter(cn -> !keyColumnNames.contains(cn)).map(JoinAddition::parse)
+                    .toArray(JoinAddition[]::new);
+            joinInput = multiJoinInputs[tableNumber] =
+                    MultiJoinInput.of(joinInput.inputTable(),
+                            joinInput.columnsToMatch(),
+                            newColumnsToAdd);
+        }
+        return joinInput;
+    }
+
+    private static Table doMultiJoinZeroKey(@NotNull final MultiJoinInput... multiJoinInputs) {
+        final MultiJoinInput[] useJoinInputs = Arrays.copyOf(multiJoinInputs, multiJoinInputs.length);
+        final TObjectIntHashMap<String> usedColumns =
+                new TObjectIntHashMap<>(multiJoinInputs[0].columnsToAdd().length, 0.5f, -1);
+
+        for (int jj = 0; jj < multiJoinInputs.length; ++jj) {
+            MultiJoinInput joinInput = useJoinInputs[jj];
+
+            if (joinInput.columnsToMatch().length != 0) {
+                Collection<String> matches = Arrays.stream(joinInput.columnsToMatch())
+                        .map(v -> v.left().name() + "=" + v.right().name()).collect(Collectors.toList());
+                throw new IllegalArgumentException(
+                        "Key column mismatch for table " + jj + ", first table had no key columns this table has "
+                                + matches);
+            }
+
+            joinInput = maybePopulateColumnsToAdd(Collections.emptySet(), useJoinInputs, jj);
```

Comment:
Maybe better to just mutate the array ourselves, rather than pass the array and an index down?